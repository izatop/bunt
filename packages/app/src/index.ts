export * from "./Payload/index.js";
export * from "./Route/index.js";
export * from "./Request/index.js";
export * from "./TransformRequest/index.js";
export * from "./Application.js";
export * from "./interfaces.js";
