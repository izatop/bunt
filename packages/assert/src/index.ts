export * from "./Exception/index.js";
export * from "./assert.js";
