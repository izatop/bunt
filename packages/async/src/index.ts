export * from "./fn.js";
export * from "./AsyncPool.js";
export * from "./AsyncState.js";
export * from "./Defer.js";
export * from "./AsyncIteratorFactory.js";
export * from "./AsyncLoader.js";
export * from "./AsyncSingleCall.js";
