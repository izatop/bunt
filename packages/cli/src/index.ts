export * from "./Transport/Headers";
export * from "./RequestCommand";
