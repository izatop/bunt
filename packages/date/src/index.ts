export * from "./DateTime.js";
export * from "./DateTimeMutation.js";
export * from "./Locale.js";
export * from "./interface.js";
