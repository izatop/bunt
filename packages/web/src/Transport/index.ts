export * from "./Response";
export * from "./Validation";
export * from "./Headers";
export * from "./interfaces";
export * from "./RequestMessage";
export * from "./Responder";
export * from "./ServerError";
export * from "./TransformError";
