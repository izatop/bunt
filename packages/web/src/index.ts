export * from "./Action/RequestHandler";
export * from "./Transport";
export * from "./WebServer";
