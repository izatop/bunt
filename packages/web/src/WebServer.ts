import * as http from "http";
import {IncomingMessage, ServerResponse} from "http";
import {Socket} from "net";
import {Application, IRoute} from "@bunt/app";
import {ActionAny, Context, ContextArg, Heartbeat, IDisposable, IRunnable, unit, Unit} from "@bunt/unit";
import {
    assert,
    Ctor,
    Logger,
    PermissionError,
    toError,
    NotFound,
    ValidationError,
    Defer,
    logger,
} from "@bunt/util";
import {IErrorResponseHeaders, IProtocolAcceptor, IResponderOptions, Responder} from "./Transport";

export class WebServer<C extends Context> extends Application<C> implements IDisposable, IRunnable {
    @logger
    declare protected readonly logger: Logger;

    readonly #server: http.Server;
    readonly #state = new Defer<void>();

    readonly #options: IResponderOptions;
    readonly #acceptors = new Map<string, IProtocolAcceptor>();
    readonly #errorCodeMap = new Map<Ctor<Error>, IErrorResponseHeaders>();

    protected constructor(unit: Unit<C>, routes: IRoute<ActionAny<C>>[] = [], options?: IResponderOptions) {
        super(unit, routes);
        this.#server = new http.Server();
        this.#options = options ?? {};

        this.#server.once("close", () => this.#state.resolve());
        this.setExceptionResponseHeaders(
            [PermissionError, {code: 403, status: "Permission denied"}],
            [ValidationError, {code: 400, status: "Validation error"}],
            [NotFound, {code: 404, status: "Not found"}],
        );
    }

    public static async factory<C extends Context>(
        context: ContextArg<C>,
        routes: IRoute<ActionAny<C>>[] = [],
        options?: IResponderOptions): Promise<WebServer<C>> {
        return new this(await unit(context), await this.preload(routes), options);
    }

    private static preload<C extends Context>(routes: IRoute<ActionAny<C>>[]): Promise<IRoute<any>[]> {
        return Promise.all(
            routes.map(async (route) => (
                Object.assign(route, {action: await Unit.getAction(route.action)})
            )),
        );
    }

    public getHeartbeat(): Heartbeat {
        return Heartbeat.create(this)
            .enqueue(this.#state)
            .onDispose(this);
    }

    public listen(port: number, backlog?: number): this {
        assert(!this.#server.listening, "Server already in listening mode");

        this.#server.on("upgrade", this.handleUpgrade);
        this.#server.on("request", this.handleRequest)
            .on("listening", () => this.logger.info("listen", {port}))
            .listen({port, backlog});

        return this;
    }

    public setUpgradeProtocolAcceptor(acceptor: IProtocolAcceptor): () => void {
        const protocol = acceptor.protocol.toLowerCase();
        assert(
            !this.#acceptors.has(protocol),
            `The ${acceptor.protocol} acceptor has already exists`,
        );

        this.logger.info("Set upgrade protocol acceptor", {protocol});
        this.#acceptors.set(protocol, acceptor);

        return () => this.#acceptors.delete(protocol);
    }

    public setExceptionResponseHeaders(...map: [error: Ctor<Error>, options: IErrorResponseHeaders][]): this {
        for (const [error, options] of map) {
            this.#errorCodeMap.set(error, options);
        }

        return this;
    }

    public async dispose(): Promise<void> {
        assert(this.#server.listening, "Server was destroyed");
        this.logger.info("close");
        this.#server.close();

        return this.#state;
    }

    protected async handle(req: IncomingMessage, res: ServerResponse): Promise<void> {
        const finish = this.logger.perf("handle", req.url);
        const request = new Responder(req, res, this.#errorCodeMap, this.#options);

        try {
            assert(request.validate(this), "Invalid Request");
            await request.respond(await this.run(request));
        } catch (error) {
            await request.respond(error);
        } finally {
            finish();
        }
    }

    private handleUpgrade = (req: IncomingMessage, socket: Socket, head: Buffer): void => {
        const {upgrade} = req.headers;
        try {
            assert(upgrade, "Upgrade headers mustn't be empty");

            const protocol = upgrade.toLowerCase();
            const acceptor = this.#acceptors.get(protocol);
            assert(acceptor, `Unsupported protocol ${protocol}`);
            acceptor.handle(req, socket, head);
        } catch (error) {
            this.logger.warning("Unexpected error", error);
            socket.write("HTTP/1.1 400 Bad request\r\nContent-Type: text/plain\r\nConnection: close\r\n\r\n");
            socket.destroy(toError(error));
        }
    };

    private handleRequest = async (req: IncomingMessage, res: ServerResponse): Promise<void> => {
        const finish = this.logger.perf("request", {method: req.method, url: req.url});

        try {
            this.logger.debug(`${req.method} ${req.url}`);
            await this.handle(req, res);
        } catch (reason) {
            this.logger.error("Handle request error", reason);

            if (!res.headersSent) {
                res.writeHead(500, "Internal Server Error");
            }
        } finally {
            if (res.writable) {
                res.end();
            }

            finish();
        }
    };
}
