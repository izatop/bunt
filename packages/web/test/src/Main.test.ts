import {JSONResponse, NoContentResponse, Redirect} from "../../src";
import * as HTTP from "http-status";

describe("Response", () => {
    test("Main", () => {
        const resp = new NoContentResponse({headers: {"foo": "123"}});
        expect(resp.getContentType()).toMatchSnapshot();
        expect(resp.getHeaders()).toMatchSnapshot();
        expect(resp.code).toBe(204);
    });

    test("JSON", () => {
        const resp = new JSONResponse({foo: "bar"}, {code: 200});
        expect(resp.getContentType()).toMatchSnapshot();
        expect(resp.getHeaders()).toMatchSnapshot();
        expect(resp.stringify()).toMatchSnapshot();
        expect(resp.code).toBe(200);
    });

    test("Redirect", () => {
        const redirect = new Redirect("/redirect");
        expect(redirect.code).toBe(HTTP["MOVED_PERMANENTLY"]);
        expect(redirect.status).toBe(HTTP["301"]);
        expect(redirect.getHeaders()).toMatchSnapshot();

        expect(() => new Redirect("/", 500)).toThrow();
    });
});
