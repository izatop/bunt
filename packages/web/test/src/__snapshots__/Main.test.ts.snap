// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response JSON 1`] = `"application/json; charset=utf-8"`;

exports[`Response JSON 2`] = `
Object {
  "content-type": "application/json; charset=utf-8",
}
`;

exports[`Response JSON 3`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\"}",
  "code": 200,
  "cookies": Array [],
  "headers": Object {
    "content-type": "application/json; charset=utf-8",
  },
  "status": "OK",
}
`;

exports[`Response Main 1`] = `"text/plain; charset=utf-8"`;

exports[`Response Main 2`] = `
Object {
  "content-type": "text/plain; charset=utf-8",
  "foo": "123",
}
`;

exports[`Response Main 3`] = `
Object {
  "body": "",
  "code": 204,
  "cookies": Array [],
  "headers": Object {
    "content-type": "text/plain; charset=utf-8",
    "foo": "123",
  },
  "status": "No Content",
}
`;

exports[`Response Redirect 1`] = `
Object {
  "body": "",
  "code": 301,
  "cookies": Array [],
  "headers": Object {
    "content-type": "text/plain; charset=utf-8",
    "location": "/redirect",
  },
  "status": "Moved Permanently",
}
`;

exports[`Response Redirect 2`] = `
Object {
  "content-type": "text/plain; charset=utf-8",
  "location": "/redirect",
}
`;
